src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs

```diff
@@ -553,6 +554,39 @@ public static MsBuildToolset GetMsBuildToolset(string userVersion, IConsole cons
             }
         }
 
+        internal static string GetNonArchitectureDirectory(string msbuildExe)
+        {
+            var msbuildFile = Path.GetFileName(msbuildExe);
+            var directory = Path.GetDirectoryName(msbuildExe);
+
+            //Given Visual Studio 2022 or later, the PATH environment variable in Developer Command Prompt contains the architecture specific path of msbuild.exe.
+            // e.g. C:\Program Files\Microsoft Visual Studio\2022\Preview\\MSBuild\Current\Bin\arm64
+            //Using the architecture specific path will cause some runtime error when loading assembly, e.g."Microsoft.Build.Framework.dll".
+            //
+            //This method is to get the non-architecture specific path of msbuild.exe if the msbuildexe is in the architecture specific folder.
+            //     C:\Program Files\Microsoft Visual Studio\2022\Preview\\MSBuild\Current\Bin\arm64
+            //  => C:\Program Files\Microsoft Visual Studio\2022\Preview\\MSBuild\Current\Bin
+            //If msbuildExe is already in the non-architecture specific folder, just return the directory.
+            foreach (var architecture in ArchitectureFolderNames)
+            {
+                if (new DirectoryInfo(directory).Name.Equals(architecture, StringComparison.OrdinalIgnoreCase))
```

Comment:
It would be better to cache the `DirectoryInfo` object so that it can used in the next line.

Reply:
Thanks for reviewing! Fixed.